### PR TITLE
Explore Metrics: Implement grouping with metric prefixes

### LINKED ISSUE #89668: Explore Metrics: Investigate grouping metrics
There are a couple of approaches we could try to display metrics. Currently, we display all of the metrics in alphabetical order. 

We can use name grouping

Image

We can use tab grouping 

Image

We can use dropdown filtering 

Image

The goal is to implement all three (maybe more in the future) in a PoC and let stakeholders explore.

### PR DESCRIPTION
**Pull request by itsmylife**
**What is this feature?**

Implementing grouping by metric names in Explore Metrics. 

Currently it looks like this 
image

Fixes: https://github.com/grafana/grafana/issues/89668

### CODE REVIEW

**Comment by catherineymgui** [comment]
Noticed this bug in the nested row option:

- Select nested row option
- Select any metric graph
- Go back to the first step in the trail
- Accordions appear horizontal

Screenshot 2024-07-03 at 09 29 50

**Review by darrenjaneczek — CHANGES_REQUESTED** [review]
Since the url parameters don't reflect the currently selected Display Mode, there is no way for the usage report to determine if these new features are being activated. Perhaps we can add another identifier under `Interactions` in `trails/interactions.ts` similar to: 

```ts
  // User changed the breakdown layout
  breakdown_layout_changed: { layout: LayoutType };
```
Something like `metric_select_display_changed`.

Check with @zhehao-grafana if it's enough for the URL parameter to identify the selected view, or if we need another item on "Interactions". Depending what he decides, we can merge after that decision is satisfied. 

**Later:**
I think there may be an opportunity to separate the different views approaches into different scenes, which use common functionality. Not urgent.

File: public/app/features/trails/groop/parser.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+      )
+    ) {
+      delimiterCount++;
+      if (delimiterCount > level) {
+        return s.slice(0, i);
+      }
+    }
+  }
+  return s; // Return the entire string if the level is higher than the number of delimiters.
+}
```
Comment (darrenjaneczek) [inline public/app/features/trails/groop/parser.ts:158]: We will also need to consider the `:` character for alert rules metrics. Any of these metrics currently group into an an empty-string group.

File: public/app/features/trails/MetricSelect/MetricSelectScene.tsx [hunk 2]
```diff
@@ -418,6 +683,22 @@ export class MetricSelectScene extends SceneObjectBase<MetricSelectSceneState> {
...
+              options={[
+                {
+                  label: 'All',
+                  value: 'all',
+                },
+                ...Array.from(rootGroup?.groups.keys() ?? []).map((g) => ({ label: g, value: g })),
+              ]}
+              value={selectedTabGroupOption}
+              onChange={model.onMetricRadioChange}
+            />
```
[the remark this answers is not in the document]
Comment (itsmylife) [reply public/app/features/trails/MetricSelect/MetricSelectScene.tsx:699]: I am not sure how I can fit that into the screen to be honest. I think RadioButtonGroup component should handle this already. Maybe I'll remove that option at all.

File: public/app/features/trails/groop/parser.ts (hunk 1 above)
Comment (itsmylife) [reply public/app/features/trails/groop/parser.ts:138]: You are right. I'll add this to my todo list. Without that, I think this feature won't be `done`.

Comment (itsmylife) [reply public/app/features/trails/groop/parser.ts:138]: Or maybe we can do it in a separate PR.

**Comment by itsmylife** [comment]
@catherineymgui I made the changes and it looks better now. Regarding to the label wording, I don't have strong opinions. But by selecting something we filter down the results based on the selection. So there is a filtering. Maybe `Filter by` can be an option.

**Comment by catherineymgui** [comment]
As mentioned in our standup, this lgtm!

File: public/app/features/trails/MetricSelect/MetricSelectScene.tsx [hunk 3]
```diff
@@ -399,6 +488,21 @@ export class MetricSelectScene extends SceneObjectBase<MetricSelectSceneState> {
...
+              onOpenMenu={() => model.reportPrefixFilterInteraction(true)}
+              onCloseMenu={() => model.reportPrefixFilterInteraction(false)}
+              options={[
+                {
+                  label: 'All metric names',
+                  value: METRIC_PREFIX_ALL,
+                },
+                ...Array.from(rootGroup?.groups.keys() ?? []).map((g) => ({ label: `${g}_`, value: g })),
+              ]}
+            />
```
Comment (bohandley) [inline public/app/features/trails/MetricSelect/MetricSelectScene.tsx:504]: Can we add a tooltip icon here to describe what we are viewing by? Something like "View by..." then some summary of this prefix docs information for metric naming convention and [a link to the docs on prefixes](https://prometheus.io/docs/practices/naming/#metric-names)

```
...should have a (single-word) application prefix relevant to the domain the metric belongs to. The prefix is sometimes referred to as namespace by client libraries. For metrics specific to an application, the prefix is usually the application name itself. Sometimes, however, metrics are more generic, like standardized metrics exported by client libraries. Examples:
  prometheus_notifications_total (specific to the Prometheus server)
  process_cpu_seconds_total (exported by many client libraries)
  http_request_duration_seconds (for all HTTP requests)
```

**Review by bohandley — COMMENTED** [review]
Looking great!!! 🥳 Just a few questions about the url sync and a request to add a tooltip to the select to add a little more context about prefixes.

File: public/app/features/trails/MetricSelect/MetricSelectScene.tsx (hunk 3 above)
Comment (catherineymgui) [reply public/app/features/trails/MetricSelect/MetricSelectScene.tsx:504]: I didn't know there was actually documentation on metric prefixes 😆 . I'm okay with this to add some more context. I believe we have to do some hacky things to make a tooltip work with a link on hover. But I feel like just text is enough?

> View by the metric prefix. A metric prefix is a single word at the beginning of the metric name, relevant to the domain the metric belongs to.

Screenshot 2024-07-19 at 15 49 08

File: public/app/features/trails/groop/parser.ts (hunk 1 above)
Comment (darrenjaneczek) [reply public/app/features/trails/groop/parser.ts:138]: Yeah. Currently the `-` does a good job of indicating "Other".

Given that the `:` character can appear anywhere (not just at the beginning), we can consider just skipping past  this character if it appears at the beginning, and using it as a prefix splitter afterward.

**Review by darrenjaneczek — APPROVED** [review]
Some suggestions in the comments to address at your convenience.
I think this is good to send out for user testing, to be refined further as noted.

**Comment by darrenjaneczek** [comment]
With this new approach, we can consider removing the unused "MetricCategory" code in
` public/app/features/trails/MetricCategory`:
- `MetricCategoryCascader.tsx`
- `useMetricCategories.ts`

**Review by bohandley — APPROVED** [review]
LGTM!

**Comment by yangkb09** [comment]
LGTM! 🚀

File: public/app/features/trails/interactions.ts
[the remark this answers is not in the document]
Comment (itsmylife) [reply public/app/features/trails/interactions.ts:1]: When I do force formatting which I do sometimes I got changes. Other than that I got nothing. So those changes were on me :)

File: public/app/features/trails/MetricSelect/MetricSelectScene.tsx (hunk 3 above)
Comment (itsmylife) [reply public/app/features/trails/MetricSelect/MetricSelectScene.tsx]: https://github.com/grafana/grafana/pull/89481/commits/f1957c57dc284a76caccb0e1cf0885c308ae2515

**Pull request merged**